File: Python/libs/RemoteApi.py
from typing import Dict, Any

from HttpCommunicationUtil import HttpCommunicationUtil
from PIL import Image

util = HttpCommunicationUtil()


class RemoteAPI:
    """
    API to simplify interation with a Surface inspector device for Remote Control
    """

    def __init__(self):
        self.util = HttpCommunicationUtil()
        self.client_secret = ""
        self.response = {}

    def set_ip(self, ip: str):
        """
        Set the Device base IP address
        :param ip: represents the scanner server
        """

        self.util.ipBase = ip

    def send_register(self, server_secret: str):
        """
         Registers the device by requesting a client secret which is later used for future commands

        :param server_secret: generated key found in the "Remote Control" feature of the Surface Inspector device
        """

        register_packet: str = self.util.construct_packet(server_secret, '', 0, '', '')
        print(self.util.ipAddressOut)
        self._set_response(util.send_packet_and_deserialize_yaml(register_packet, self.util.ipAddressOut))
        self.client_secret = self.get_response_msg()

    def send_start_scan(self):
        """ Sends the request to Start scanning. """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'START_SCAN', '')
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def send_stop_scan(self):
        """ Sends the request to Stop scanning. """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'STOP_SCAN', '')
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def send_get_state(self):
        """ Requests the current scanner State. """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'GET_STATE', '')
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def send_set_nickname(self, name: str):
        """
        Sets the nickname for the most recent scan/photo. The name is set in 'value'.

        :param name: a String to name the Scan file
        """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'SET_NICKNAME', name)
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def send_capture_image_frame(self):
        """
        Captures a video frame from the video stream, returning a BufferedImage for further manipulation

        :return: PIL Image object for further interaction/manipulation
        """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'CAPTURE_VIDEO_FRAME', '')
        return util.send_packet_and_deserialize_image(command_packet, self.util.ipAddressOut)

    def send_upload_scan(self, file_name: str):
        """
        Uploads the Scan to Ninox360 Cloud storage. The file is set in 'value'.

        :param file_name: which we are uploading
        """

        command_packet = self.util.construct_packet('', self.client_secret, 0, 'UPLOAD_REMOTE', file_name)
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def send_get_file(self, file_name: str):
        """
        Request that the scanner send over the specified file.

        :param file_name: which we are requesting
        """
        command_packet = self.util.construct_packet('', self.client_secret, 0, 'GET_FILE', file_name)
        self._set_response(util.send_packet_and_deserialize_yaml(command_packet, self.util.ipAddressOut))

    def _set_response(self, response: Dict[str, Any]):
        """ sets the server response as a map """
        self.response = response

    def get_response_msg(self) -> str:
        """ :return the server response message from the previous command issued """
        return self.response.get('message')

    def get_response_status(self) -> str:
        """ :return the server response status from the previous commmand issued """
        return self.response.get('response')

    @staticmethod
    def save_image_to_disk(image: Image, file_name: str = 'image.jpg'):
        """
        Utility to save a file locally to disk

        :param image: PIL Image data which we are saving
        :param file_name: for the file which we are saving
        """

        with open(file_name, 'wb') as image_file:
            image.save(file_name)

    def disconnect(self):
        """ Resets the client_secret to empty requiring another registration for future commands """
        self.client_secret = ''

File: Java/libs/src/main/java/ninox360/util/RemoteAPI.java
package ninox360.util;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Map;

/**
 * API to simplify interation with a Surface inspector device for Remote Control.
 */
public class RemoteAPI {
    HttpCommunicationUtil util = new HttpCommunicationUtil();
    String clientSecret;
    PacketRemoteResponse responsePacket = new PacketRemoteResponse();

    /**
     * Set the Device base IP address
     *
     * @param ip address representing the scanner server
     */
    public void setIP(String ip) {
        HttpCommunicationUtil.ipBase = ip;
    }

    /**
     * Sets the local response variables based on the deserialized response map
     *
     * @param response Map of the Response and Message from PacketRemoteResponse
     */
    private void setResponse(Map<String, Object> response) {
        responsePacket.response = (String) response.get("response");
        responsePacket.message = (String) response.get("message");
    }

    /**
     * Indicates a status after an outbound command/register packet has been sent, stored until the next packet is sent.
     * <pre>
     *     Usage Example:
     *     api.sendRegister(serverSecret);
     *
     *     String result = api.getResponseStatus();
     *     System.out.println("Response: " + result); // Prints the response status code from outgoing the request
     *
     *     // sample out:  response: SUCCESS
     * </pre>
     *
     * @return returns the corresponding network status string from a prior response.
     */
    public String getResponseStatus() {
        return responsePacket.response;
    }

    /**
     * Gets the server response message following a request, stored until the next packet is sent.
     * <pre>
     *     Usage Example:
     *     api.sendRegister(serverSecret);
     *
     *     String result = api.getResponseMsg();
     *     System.out.println("Message: " + result); // printing out the client secret returned
     *
     *     // sample out:  Message: BLZ1x
     * </pre>
     *
     * @return returns the corresponding message string, in the example: "MaDA1" would be returned
     */
    public String getResponseMsg() {
        return responsePacket.message;
    }

    /**
     * Registers the device by requesting a client secret which is later used for future commands
     *
     * @param serverSecret Generated key found in the "Remote Control" feature of the Surface Inspector device
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendRegister(String serverSecret) throws IOException {
        String registerPacket = HttpCommunicationUtil.constructPacket(serverSecret, "", 0, "", "");
        setResponse(util.sendPacketAndDeserializeYaml(registerPacket, HttpCommunicationUtil.ipAddressOut));

        // set the client secret to the response message for future commands
        this.clientSecret = getResponseMsg();
    }

    /**
     * Sends the request to Start scanning.
     *
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendStartScan() throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "START_SCAN", "");
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sends the request to Stop scanning.
     *
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendStopScan() throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "STOP_SCAN", "");
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Requests the current scanner State.
     *
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendGetState() throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "GET_STATE", "");
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sets the camera stillness requirement.
     *
     * @param state: Boolean indicating if the stillness is required prior to scanning, true=on false=off
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendStillnessRequired(boolean state) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "REQUIRE_STILL", Boolean.toString(state));
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sets the camera QR marker requirement.
     *
     * @param state: Boolean indicating if the QR marker is required prior to scanning, true=on false=off
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendMarkerRequired(boolean state) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "REQUIRE_MARKER", Boolean.toString(state));
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sets the camera sensor exposure time (nanoseconds) aka shutter speed.
     *
     * @param exposure_ns, exposure time aka shutter speed, represented in nanoseconds
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendSetCameraExposure(long exposure_ns) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "EXPOSURE", Long.toString(exposure_ns));
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sets the camera sensor sensitivity aka ISO.
     *
     * @param sensitivity_iso, exposure time aka shutter speed, represented in nanoseconds
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendSetCameraSensitivity(int sensitivity_iso) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "SENSITIVITY", Integer.toString(sensitivity_iso));
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Sets the nickname for the most recent scan/photo. The name is set in 'value'.
     *
     * @param name a String to name the Scan file
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendSetNickname(String name) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "SET_NICKNAME", name);
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Captures a video frame from the video stream, returning a BufferedImage for further manipulation
     *
     * @return BufferedImage for further interaction/manipulation
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public BufferedImage sendCaptureImageFrame() throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "CAPTURE_VIDEO_FRAME", "");
        return util.sendPacketAndDeserializeImage(commandPacket, HttpCommunicationUtil.ipAddressOut);
    }

    /**
     * Uploads the Scan or File to Ninox360 Cloud storage.
     *
     * @param fileName of the file which we are uploading
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendUploadScan(String fileName) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "UPLOAD_REMOTE", fileName);
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Request that the scanner send over the specified file.
     *
     * @param fileName Filename which we are getting from the server.
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process
     */
    public void sendGetFile(String fileName) throws IOException {
        String commandPacket = HttpCommunicationUtil.constructPacket("", clientSecret, 0, "GET_FILE", fileName);
        setResponse(util.sendPacketAndDeserializeYaml(commandPacket, HttpCommunicationUtil.ipAddressOut));
    }

    /**
     * Disconnect this client from the scanner. Forget the client specific secret.
     */
    public void disconnect() {
        clientSecret = "";
    }

    /**
     * API utility for saving images to disk as a JPG.
     *
     * @param image    BufferedImage object
     * @param filePath storage path
     * @return the boolean result from the ImageIO.write() call
     * @throws IOException An exception will be thrown if any process fails,
     *                     detailed information will be displayed depending on the process + The file data
     */
    public static boolean saveImageToDisk(BufferedImage image, String filePath) throws IOException {
        File file = new File(filePath);
        try {
            System.out.println("Saving image to disk: " + filePath + " at path: " + file.getAbsolutePath());
            return ImageIO.write(image, "jpg", file);
        } catch (Exception e) {
            throw new IOException("Failed to save the image: " + filePath + " at path: " + file.getAbsolutePath());
        }
    }
}
